File: 2019/day04.py

def passes1(x):
	s = str(x)
	inc = True
	rep = False
	for i in range(1,len(s)):
		if (s[i] < s[i-1]):
			inc = False
			break
		if (s[i-1] == s[i]):
			rep = True
	return (inc and rep)

def passes2(x):
	s = str(x)
	inc = True
	rep = False
	for i in range(1,len(s)):
		if (s[i] < s[i-1]):
			inc = False
			break
		if (s[i-1] == s[i]):
			multiple = False
			for j in range(i-1):
				if (s[j]==s[i]):
					multiple = True
					break
			if (multiple==False):
				for j in range(i+1, len(s)):
					if (s[j]==s[i]):
						multiple = True
						break
			if (multiple==False):
				rep = True
	return (inc and rep)

min = 265275
max = 781584
count1 = 0
count2 = 0
for i in range(min,max+1):
	count1 = count1 + passes1(i)
	count2 = count2 + passes2(i)
print("Part 1: answer is {:d}".format(count1))
print("Part 2: answer is {:d}".format(count2))

File: 2019/day04.c
#include <stdio.h>
#include <string.h>

int passes1(int x) {

	char s[7];
	sprintf(s, "%d", x);
	int inc = 1;
	int rep = 0;
	for (int i=1; i<strlen(s); i++) {
		if (s[i] < s[i-1]) {
			inc = 0;
			break;
		}
		if (s[i-1] == s[i]) {
			rep = 1;
		}
	}
	return(inc & rep);
}

int passes2(int x) {

	char s[7];
	sprintf(s, "%d", x);
	int inc = 1;
	int rep = 0;
	for (int i=1; i<strlen(s); i++) {
		if (s[i] < s[i-1]) {
			inc = 0;
			break;
		}
		if (s[i-1] == s[i]) {
			int multiple = 0;
			for (int j=0; j<(i-1); j++) {
				if (s[j]==s[i]) {
					multiple = 1;
					break;
				}
			}
			if (multiple==0) {
				for (int j=i+1; j<strlen(s); j++) {
					if (s[j]==s[i]) {
						multiple = 1;
						break;
					}
				}
			}
			if (multiple==0) { rep = 1; }
		}
	}
	return(inc & rep);
}

int main(int argc, char *argv[]) {

	int min = 265275;
	int max = 781584;
	int count1 = 0;
	int count2 = 0;
	for (int i=min; i<=max; i++) {
		count1 = count1 + passes1(i);
		count2 = count2 + passes2(i);
	}
	printf("Part 1: answer is %d\n", count1);
	printf("Part 2: answer is %d\n", count2);

	return 0;
}